<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SNP Effect Matrix Pipeline in R • SEMplR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="SNP Effect Matrix Pipeline in R">
<meta name="description" content="SEMplR computes transcription factor binding affinity scores for genomic positions and genetic variants. Scores are computed from SNP Effect Matrices (SEMs) produced by SEMpl. 223 pre-computed SEMs are included with the package or custom sets can be provided. Enrichment can be tested among sets of genomic positions to determine if transcription factor binding events occur more often than expected. Comparing binding affinity scores between alleles can reveal differences in transcription factor binding with genetic variation. This package also includes several visualization functions to view scores both on the motif and variant/position level.">
<meta property="og:description" content="SEMplR computes transcription factor binding affinity scores for genomic positions and genetic variants. Scores are computed from SNP Effect Matrices (SEMs) produced by SEMpl. 223 pre-computed SEMs are included with the package or custom sets can be provided. Enrichment can be tested among sets of genomic positions to determine if transcription factor binding events occur more often than expected. Comparing binding affinity scores between alleles can reveal differences in transcription factor binding with genetic variation. This package also includes several visualization functions to view scores both on the motif and variant/position level.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SEMplR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/SEMplR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/grkenney/SEMplR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="semplr-">SEMplR <a href="https://grkenney.github.io/SEMplR"><img src="reference/figures/SEMplR-B.png" align="right" height="200" alt="SEMplR website" style="float:right; height:200px;"></a>
<a class="anchor" aria-label="anchor" href="#semplr-"></a>
</h1></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>SEMplR (SNP Effect Matrix Pipeline in R) is an R package that predicts transcription factor (TF) binding. SEMplR can be used to predict binding affinity of TFs at genomic loci or predict the affect of genetic variation on TF binding.</p>
<p>SEMplR scores genomic regions or sequences of interest against SNP Effect Matrices (SEMs). SEMs are position x nucleotide matrix, generated by integrating information from position weighted matrices (PWMs), ChIP-seq, and DNase-seq data. This integration of binding data means that motif analysis with SEMs is more indicative of true binding potential compared to traditional motif analyses with PWMs where scores are more indicative of sequence similarity to consensus motifs. You can read more about SEMs and how they are generated in the original <a href="https://doi.org/10.1093/bioinformatics/btz612" class="external-link">SEMpl paper</a>.</p>
<p>This package extends the functionality of the <a href="https://github.com/Boyle-Lab/SEMpl" class="external-link">SEMpl</a> (SNP Effect Matrix pipeline) command line tool developed by the Boyle Lab at the University of Michigan. To support data analysis and visualizations with SEMs.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use SEMplR in your work, please also cite SEMpl:</p>
<p>Sierra S Nishizaki, Natalie Ng, Shengcheng Dong, Robert S Porter, Cody Morterud, Colten Williams, Courtney Asman, Jessica A Switzenberg, Alan P Boyle, Predicting the effects of SNPs on transcription factor binding affinity, <em>Bioinformatics</em>, Volume 36, Issue 2, 15 January 2020, Pages 364–372, <a href="https://doi.org/10.1093/bioinformatics/btz612" class="external-link uri">https://doi.org/10.1093/bioinformatics/btz612</a></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<pre><code><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"grkenney/SEMplR"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>Below are some examples of basic usage. Please see the <a href="https://grkenney.github.io/SEMplR/">vignette</a> for more detailed workflow examples.</p>
<div class="section level3">
<h3 id="predicting-transcription-factor-binding">Predicting transcription factor binding<a class="anchor" aria-label="anchor" href="#predicting-transcription-factor-binding"></a>
</h3>
<p>SEMplR accepts GRanges objects or lists of sequences to score. Here, we analyze two loci with SEMplR’s default set of 223 pre-computed SEMs, stored in the <code>SEMC</code> object. The <code>scoreBinding</code> function produces a data object with information about the ranges analyzed, SEM meta data, and a table with 446 rows (an entry for each loci and SEM combination).</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the default set of SEMs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">SEMC</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define genomic loci to score</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr12"</span>, <span class="st">"chr19"</span><span class="op">)</span>,</span>
<span>                             ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">94136009</span>, <span class="fl">10640062</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># score TF binding at each loci</span></span>
<span><span class="va">sb</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scoreBinding.html">scoreBinding</a></span><span class="op">(</span><span class="va">gr</span>,</span>
<span>                   sem <span class="op">=</span> <span class="va">SEMC</span>,</span>
<span>                   genome <span class="op">=</span> <span class="va">Hsapiens</span><span class="op">)</span></span></code></pre>
<p>When analyzing large sets of loci, it can be helpful to know if one or more TFs are bound more than we would expect by chance. SEMplR includes enrichment and plotting functions to address this question.</p>
<pre><code><span><span class="co"># compute enrichment</span></span>
<span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrichSEMs.html">enrichSEMs</a></span><span class="op">(</span><span class="va">sb</span>, <span class="va">SEMC</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot enrichment results</span></span>
<span><span class="fu"><a href="reference/plotEnrich.html">plotEnrich</a></span><span class="op">(</span><span class="va">e</span>, <span class="va">SEMC</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="predicting-effect-of-genetic-variation-on-transcription-factor-binding">Predicting effect of genetic variation on transcription factor binding<a class="anchor" aria-label="anchor" href="#predicting-effect-of-genetic-variation-on-transcription-factor-binding"></a>
</h3>
<p>SEMplR accepts both VRanges and GRanges objects, specifying a reference an alternative allele. Every variant is scored against every SEM and a scoring is done for each allele independently.</p>
<p>The resulting object contains three slots containing the variants scored, SEM meta data, and the scoring table. These can be accessed with the <code>variants()</code>, <code><a href="reference/semData.html">semData()</a></code>, and <code><a href="reference/scores.html">scores()</a></code> functions respectively.</p>
<pre><code><span><span class="va">vr</span> <span class="op">&lt;-</span> <span class="fu">VRanges</span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr12"</span>, <span class="st">"chr19"</span><span class="op">)</span>,</span>
<span>              ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">94136009</span>, <span class="fl">10640062</span><span class="op">)</span>,</span>
<span>              ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>, <span class="st">"T"</span><span class="op">)</span>, alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="st">"A"</span><span class="op">)</span>,</span>
<span>              id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scoreVariants.html">scoreVariants</a></span><span class="op">(</span>vr <span class="op">=</span> <span class="va">vr</span>,</span>
<span>                    sem <span class="op">=</span> <span class="va">SEMC</span>,</span>
<span>                    genome <span class="op">=</span> <span class="va">Hsapiens</span><span class="op">)</span></span></code></pre>
<p>SEMplR includes two plotting functions to help users predict (1) which TFs change binding with a genetic variant and (2) which variants change the binding of a TF.</p>
<pre><code><span><span class="fu"><a href="reference/plotSemVariants.html">plotSemVariants</a></span><span class="op">(</span><span class="va">s</span>, <span class="st">"IKZF1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plotSemMotifs.html">plotSemMotifs</a></span><span class="op">(</span><span class="va">s</span>, <span class="st">"A"</span><span class="op">)</span></span></code></pre>
<p>Please see more information on these plots and their interpretation in our <a href="https://grkenney.github.io/SEMplR/">vignette</a>.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/grkenney/SEMplR/" class="external-link">Browse source code</a></li>
<li><a href="https://www.github.com/grkenney/SEMplR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SEMplR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/grkenney" class="external-link">Grace Kenney</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0009-6308-3150" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Douglas Phanstiel <br><small class="roles"> Author </small>   </li>
<li>NIH NIGMS <br><small class="roles"> Funder </small>   </li>
<li>NSF GRFP <br><small class="roles"> Funder </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/grkenney" class="external-link">Grace Kenney</a>, Douglas Phanstiel, NIH NIGMS, NSF GRFP.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
