---
title: "SEMplR_Vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SEMplR_Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>", 
  fig.height = 5,
  fig.width = 5
)
```

```{r load_pkgs, message=FALSE}
library(SEMplR)
library(VariantAnnotation)
```

# Setup
First, we will create a `VRanges` object to store our variants of interest.
```{r init_vr}
vr <- VRanges(seqnames = c("chr12", "chr19"),
              ranges = c(94136009, 10640062), 
              ref = c("G", "T"), alt = c("C", "A"))
vr$id <- 1:length(vr)
```

# Loading SEMs
Next, we will load our SNP Effect Matrices and baseline file. Note that this process may take a couple minutes, if files are downloaded locally, use the `sem_dir` and `baseline_file` options for a faster run time.
```{r get_files}
sems <- loadSEMs(semDir = "SEMs",
                 baselineFile = "SEMs/BASELINE/SEMs_baseline_norm.txt",
                 metadataFile =  "metadata.csv")
sems[2]
```

```{r}
plotMotif(sems, "MA0099.2_HeLa")
```


# Scoring Motifs
Next, we will score the reference and alternate alleles for each variant against the SEMs and normalize to the baseline for each motif.
```{r score_motifs}
sempl_obj <- semMotifBinding(vr = vr, semList = sems)
sempl_obj
```

The resulting object contains two slots:

- `variants`: The `VRanges` object for each 
- `scores`: A data.table object that . There are X columns:
    - **seqnames** - chromosome name
    - **ranges** - position of variant
    - **sem** - SEM name
    - **nonRiskSeq** - reference allele frame with highest binding score
    - **riskSeq** - alternate allele frame with highest binding score
    - **nonRiskScore** - score of reference allele for given SEM
    - **riskScore** - score of alternate allele for given SEM
    - **nonRiskNorm** - normalized score of reference allele for given SEM
    - **riskNorm** - normalized score of alternate allele for given SEM

# Visualization

```{r plot_motif1}
plotSemMotifs(sempl_obj, variant = "1", label="tf")
```

```{r plot_motif2}
plotSemMotifs(sempl_obj, variant="1")
```
```{r}
plotSemVariants(sempl_obj, "M00496")
```

# Enrichment
To test for enrichment of differential binding propensity among multiple variants.
```{r}
# load our example variants
vars_df <- utils::read.delim("example_vars.csv", sep=",")[1:50, -1]

# create vrange
vr <- VariantAnnotation::VRanges(seqnames = vars_df$seqnames,
                                 ranges = IRanges::IRanges(start=vars_df$start,
                                                           end=vars_df$end),
                                 ref = vars_df$ref,
                                 alt = vars_df$alt)
# add the rsid to the metadata
vr$id <- vars_df$id
```

```{r}
# score the variants
sempl_obj <- semMotifBinding(vr = vr, semList = sems)
```


```{r}
# calculate the enrichment score
e <- enrichment(sempl_obj, d = "gained")
e
```
```{r}
plotSemVariants(sempl_obj, sem_id = "M01266")
```

```{r}
var_scores <- scores(sempl_obj)[varId == "rs61943586" &
                                  semId == "M01266"]

plotMotif(sems, "M01266", 
          refSeq = var_scores$nonRiskSeq, 
          altSeq = var_scores$riskSeq,
          refIndex = var_scores$nonRiskVarIndex,
          altIndex = var_scores$riskVarIndex)
```


```{r}
var_scores <- scores(sempl_obj)[varId == "rs4318227" &
                                  semId == "M01266"]

plotMotif(sems, "M01266", 
          refSeq = var_scores$nonRiskSeq, 
          altSeq = var_scores$riskSeq,
          refIndex = var_scores$nonRiskVarIndex,
          altIndex = var_scores$riskVarIndex)
```

```{r}
plotMotif(sems, "M01266", 
          refSeq = var_scores$nonRiskSeq, 
          refIndex = var_scores$nonRiskVarIndex)
```

```{r}
plotMotif(sems, "M01266", 
          altSeq = var_scores$riskSeq,
          altIndex = var_scores$riskVarIndex)
```

