---
title: "SEMplR Vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SEMplR_Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>", 
  fig.height = 5,
  fig.width = 5
)
```

```{r}
suppressPackageStartupMessages({
  library(VariantAnnotation)
  library(SEMplR)
})
```


Printing the `SNPEffectMatrixCollection`, we can see that we have 223 SEMs and 13 meta data features for each.
```{r}
data(sc)
sc
```


We can access all SEMs in collection the with the function `sems()` or some subset of matrices by specifying a vector of semIds in the `semId` parameter.
```{r}
sems(sc, semId = c("AP2B_HUMAN.SK-N-SH", "ZSCAN4_secondary"))
```

We can view the SEM meta data slot with the `semData` function.
```{r}
semData(sc)
```

# Build Variant Set

We will load the variants we want to score as a `VRanges` object.
```{r}
vr <- VRanges(seqnames = c("chr12", "chr19"),
              ranges = c(94136009, 10640062), 
              ref = c("G", "T"), alt = c("C", "A"))
vr
```

# Score Variants
```{r}
sempl_obj <- scoreVariants(vr = vr, semList = sc, bs_genome_obj=BSgenome.Hsapiens.UCSC.hg38::Hsapiens)
sempl_obj
```

```{r}
variants(sempl_obj)
```

```{r}
semData(sempl_obj)
```

```{r}
scores(sempl_obj)
```


# View Frame
```{r}
viewFrames(sempl_obj, vid = "chr12:94136009:G>C", sid = "AP2B_HUMAN.SK-N-SH")
```

```{r}
plotSEM(sc, motif = "AP2B_HUMAN.SK-N-SH", 
        motifSeq = "GCTTTGAGGC", highlight = 1)
plotSEM(sc, motif = "AP2B_HUMAN.SK-N-SH", 
        motifSeq = "GCTTTCAGGC", highlight = 1, hcol = "red")
```

```{r}
plotSemMotifs(sempl_obj, variant = "chr19:10640062:T>A")
```

```{r}
plotSemMotifs(sempl_obj, variant = "chr19:10640062:T>A",
              label = "transcription_factor")
```


```{r}
plotSemVariants(sempl_obj, semId = "TFAP4_HUMAN.HepG2")
```

# Enrichment

```{r}
# loading data
caqtl_file <- "/Users/gracekenney/Downloads/caQTL_variants_overlappingPeaks_LD-r2-0.8_withLead.bed"
caqtls <- data.table::fread(file = caqtl_file, sep = "\t")[1:1000,]
proxy_ID_split <- strsplit(caqtls$proxy_ID, ":")

# formatting position and alleles
chr <- lapply(proxy_ID_split, `[[`, 1) |> unlist()
pos <- lapply(proxy_ID_split, `[[`, 2) |> unlist()
ref_allele <- lapply(proxy_ID_split, `[[`, 3) |> unlist()
alt_allele <- lapply(proxy_ID_split, `[[`, 4) |> unlist()

# building variant annotation obj
vr_e <- VariantAnnotation::VRanges(seqnames = chr,
                                 ranges = pos, 
                                 ref = ref_allele, 
                                 alt = alt_allele)
data(sc)

# score variants
var_scores_e <- scoreVariants(vr = vr_e, 
                            semList = sc,
                            bs_genome_obj=BSgenome.Hsapiens.UCSC.hg38::Hsapiens)

plotSemVariants(var_scores_e, semId = "MA0151.1")
```

```{r}
plotSemMotifs(var_scores_e, "chr1:906982:C>T", label = "transcription_factor")
```

# Binding Enrichment

Given this is a large set of variants, we might want to check if some 
transcription factors are enriched for binding in one allele or the other.

Here, we will test if any of these transcription factors are bound more than
expected in the reference alleles of this set.

```{r}
gr_e <- GenomicRanges::GRanges(seqnames = chr,
                                 ranges = pos, 
                                 allele = ref_allele)

e <- enrichSEMs(x = gr_e, semList = sc, 
                bs_genome_obj = BSgenome.Hsapiens.UCSC.hg38::Hsapiens)
```

```{r}
plt <- plotEnrich(e, semList = sc)
plt
```


# Extras

# Loading SEMs

First, we create a list of file paths to the .sem files we want to include. We will also load the meta data for each sem in a `data.table` object. If meta data is used, all SEMs must be represented in the meta data table.
```{r, eval=F}
sem_files <- list.files("SEMs/", 
                        pattern = ".sem", full.names = T)
metadata_file <- "SEMs/sempl_metadata.csv"
meta <- data.table::fread(file = metadata_file, sep = ",")

head(meta)
```

We will load the matrix and meta data for all SEMs in a single `SNPEffectMatrixCollection` object. This object has two slots, one containing a named list of the matrices and a second slot containing our meta data table with a key column connecting the meta data to the names of the matrices.
```{r, eval=F}
sc <- loadSEMCollection(semFiles = sem_files, 
                        semMetaData = meta, 
                        semMetaKey = "SEM")
sc
```
