<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Retrieve Promoter Regions and Sample Background Elements — getCoordinates • SEMplR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Retrieve Promoter Regions and Sample Background Elements — getCoordinates"><meta name="description" content="Given a mapped set of foreground IDs (with Entrez and mappedID),
this function:
Fetches genomic coordinates for each feature (gene or transcript),
according to the chosen TSS.method.
Extracts promoter windows around those coordinates.
Optionally reduces overlaps and selects one promoter per gene.
Samples background promoter elements matching the foreground

"><meta property="og:description" content="Given a mapped set of foreground IDs (with Entrez and mappedID),
this function:
Fetches genomic coordinates for each feature (gene or transcript),
according to the chosen TSS.method.
Extracts promoter windows around those coordinates.
Optionally reduces overlaps and selects one promoter per gene.
Samples background promoter elements matching the foreground

"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SEMplR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SEMplR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/grkenney/SEMplR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Retrieve Promoter Regions and Sample Background Elements</h1>
      <small class="dont-index">Source: <a href="https://github.com/grkenney/SEMplR/blob/HEAD/R/getCoordinates.R" class="external-link"><code>R/getCoordinates.R</code></a></small>
      <div class="d-none name"><code>getCoordinates.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given a mapped set of foreground IDs (with Entrez and mappedID),
this function:</p><ol><li><p>Fetches genomic coordinates for each feature (gene or transcript),
according to the chosen <code>TSS.method</code>.</p></li>
<li><p>Extracts promoter windows around those coordinates.</p></li>
<li><p>Optionally reduces overlaps and selects one promoter per gene.</p></li>
<li><p>Samples background promoter elements matching the foreground</p></li>
</ol></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getCoordinates</span><span class="op">(</span></span>
<span>  <span class="va">mapped</span>,</span>
<span>  <span class="va">txdb</span>,</span>
<span>  transcript <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  promoterWindow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>upstream <span class="op">=</span> <span class="fl">300</span>, downstream <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>  standardChroms <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reduceOverlaps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  overlapMinGap <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  onePromoterPerGene <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mapped">mapped<a class="anchor" aria-label="anchor" href="#arg-mapped"></a></dt>
<dd><p>A list as returned by <code><a href="mapIDs.html">mapIDs</a>()</code>, containing
at least <code>fg_ids</code> and <code>bg_ids</code>.</p></dd>


<dt id="arg-txdb">txdb<a class="anchor" aria-label="anchor" href="#arg-txdb"></a></dt>
<dd><p>A TxDb object.
(e.g. <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code>).</p></dd>


<dt id="arg-transcript">transcript<a class="anchor" aria-label="anchor" href="#arg-transcript"></a></dt>
<dd><p>Logical; <code>TRUE</code> for transcript-level coordinates,
<code>FALSE</code> for gene-level.</p></dd>


<dt id="arg-n-ratio">n_ratio<a class="anchor" aria-label="anchor" href="#arg-n-ratio"></a></dt>
<dd><p>Numeric; ratio of ranges to retain in the background set.
The number of background ranges will be equal to <code>n_ratio</code> multiplied
by the number of foreground ranges.</p></dd>


<dt id="arg-promoterwindow">promoterWindow<a class="anchor" aria-label="anchor" href="#arg-promoterwindow"></a></dt>
<dd><p>Numeric named vector of lengths:
<code>c(upstream, downstream)</code> (default <code>c(300,50)</code>).</p></dd>


<dt id="arg-standardchroms">standardChroms<a class="anchor" aria-label="anchor" href="#arg-standardchroms"></a></dt>
<dd><p>Logical; restrict to standard chromosomes.</p></dd>


<dt id="arg-reduceoverlaps">reduceOverlaps<a class="anchor" aria-label="anchor" href="#arg-reduceoverlaps"></a></dt>
<dd><p>Logical; merge any overlapping promoter windows.</p></dd>


<dt id="arg-overlapmingap">overlapMinGap<a class="anchor" aria-label="anchor" href="#arg-overlapmingap"></a></dt>
<dd><p>Numeric; minimum gap when reducing overlaps.</p></dd>


<dt id="arg-onepromoterpergene">onePromoterPerGene<a class="anchor" aria-label="anchor" href="#arg-onepromoterpergene"></a></dt>
<dd><p>Logical; if <code>TRUE</code>, choose one
promoter per gene.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A named <code>list</code> from the final
<code>.defineBackgroundElements()</code>:</p>
<dl><dt>backgroundElements</dt>
<dd><p><code>GRanges</code> of sampled background promoters</p></dd>

<dt>foregroundElements</dt>
<dd><p><code>GRanges</code> of foreground promoters</p></dd>

<dt>backgroundUniverse</dt>
<dd><p><code>GRanges</code> of the pruned universe</p></dd>

<dt>matchObject</dt>
<dd><p><code>MatchedGRanges</code> if <code>bgMethod="matched"</code>,
else <code>NULL</code></p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span></span></span>
<span class="r-in"><span><span class="va">orgdb</span> <span class="op">&lt;-</span> <span class="va">org.Hs.eg.db</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">my_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000139618"</span>, <span class="st">"ENSG00000157764"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="mapIDs.html">mapIDs</a></span><span class="op">(</span>orgdb <span class="op">=</span> <span class="va">orgdb</span>, </span></span>
<span class="r-in"><span>              foreground_ids <span class="op">=</span> <span class="va">my_genes</span>, </span></span>
<span class="r-in"><span>              id_type <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in mapIDs(orgdb = orgdb, foreground_ids = my_genes, id_type = "ENSEMBL"):</span> could not find function "mapIDs"</span>
<span class="r-in"><span><span class="va">filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="poolFilter.html">poolFilter</a></span><span class="op">(</span><span class="va">ids</span>, geneType<span class="op">=</span><span class="st">"protein-coding"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in poolFilter(ids, geneType = "protein-coding"):</span> could not find function "poolFilter"</span>
<span class="r-in"><span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu">getCoordinates</span><span class="op">(</span>mapped <span class="op">=</span> <span class="va">filtered</span>, txdb <span class="op">=</span> <span class="va">txdb</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in getCoordinates(mapped = filtered, txdb = txdb):</span> could not find function "getCoordinates"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/grkenney" class="external-link">Grace Kenney</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

