<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get sequence for genomic ranges and variants — getRangeSeqs • SEMplR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Get sequence for genomic ranges and variants — getRangeSeqs"><meta name="description" content="Query sequences associated with a given range in a reference genome and
optionally include nucleotides up and downstream of the range of interest
sequences. Store the results in the metadata of the range object."><meta property="og:description" content="Query sequences associated with a given range in a reference genome and
optionally include nucleotides up and downstream of the range of interest
sequences. Store the results in the metadata of the range object."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SEMplR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SEMplR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/grkenney/SEMplR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get sequence for genomic ranges and variants</h1>
      <small class="dont-index">Source: <a href="https://github.com/grkenney/SEMplR/blob/devel/R/getRangeSeqs.R" class="external-link"><code>R/getRangeSeqs.R</code></a></small>
      <div class="d-none name"><code>getRangeSeqs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Query sequences associated with a given range in a reference genome and
optionally include nucleotides up and downstream of the range of interest
sequences. Store the results in the metadata of the range object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getRangeSeqs</span><span class="op">(</span><span class="va">x</span>, <span class="va">genome</span>, up <span class="op">=</span> <span class="fl">0</span>, down <span class="op">=</span> <span class="fl">0</span>, refCol <span class="op">=</span> <span class="cn">NULL</span>, altCol <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code>VRanges</code> or <code>GRanges</code> object with one or more variants. seqnames
and ranges fields are required.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>A <code>BSgenome</code> object for the genome build to use.</p></dd>


<dt id="arg-up">up<a class="anchor" aria-label="anchor" href="#arg-up"></a></dt>
<dd><p>Numeric, number of bases to return upstream of variant</p></dd>


<dt id="arg-down">down<a class="anchor" aria-label="anchor" href="#arg-down"></a></dt>
<dd><p>Numeric, number of bases to return downstream of variant</p></dd>


<dt id="arg-refcol">refCol<a class="anchor" aria-label="anchor" href="#arg-refcol"></a></dt>
<dd><p>Column name in meta data storing the ref allele</p></dd>


<dt id="arg-altcol">altCol<a class="anchor" aria-label="anchor" href="#arg-altcol"></a></dt>
<dd><p>Column name in meta data storing the alt allele</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code>x</code> with added meta data columns. By default, only a "sequence"
column is added to the meta data. If variant information is supplied,
either in a <code>VRanges</code> object or through the "allele" parameter, "ref_seq"
and "alt_seq" columns are added to the meta data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># set reference genome</span></span></span>
<span class="r-in"><span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">Hsapiens</a></span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Query sequnece for a GRanges object</span></span></span>
<span class="r-in"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    seqnames <span class="op">=</span> <span class="st">"chr12"</span>,</span></span>
<span class="r-in"><span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">94136009</span>, <span class="fl">94136019</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">getRangeSeqs</span><span class="op">(</span><span class="va">gr</span>, genome <span class="op">=</span> <span class="va">b</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 1 metadata column:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames            ranges strand |    sequence</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]    chr12 94136009-94136019      * | GAGGCATCTGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Query variant sequences for a VRanges object with 10bp flanks</span></span></span>
<span class="r-in"><span><span class="va">vr</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">VRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    seqnames <span class="op">=</span> <span class="st">"chr12"</span>,</span></span>
<span class="r-in"><span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">94136009</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    ref <span class="op">=</span> <span class="st">"G"</span>, alt <span class="op">=</span> <span class="st">"C"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">getRangeSeqs</span><span class="op">(</span><span class="va">vr</span>, genome <span class="op">=</span> <span class="va">b</span>, up <span class="op">=</span> <span class="fl">10</span>, down <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> VRanges object with 1 range and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand         ref              alt     totalDepth</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; &lt;character&gt; &lt;characterOrRle&gt; &lt;integerOrRle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]    chr12  94136009      *           G                C           &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             refDepth       altDepth   sampleNames softFilterMatrix |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       &lt;integerOrRle&gt; &lt;integerOrRle&gt; &lt;factorOrRle&gt;         &lt;matrix&gt; |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]           &lt;NA&gt;           &lt;NA&gt;          &lt;NA&gt;                  |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     ref_seq               alt_seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 &lt;character&gt;           &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1] AGAAGGCTTTGAGGCATCTGC AGAAGGCTTTCAGGCATCTGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   hardFilters: NULL</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/grkenney" class="external-link">Grace Kenney</a>, Douglas Phanstiel, NIH NIGMS, NSF GRFP.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

