<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate binding propensity for all SEM motifs and genomic positions provided — scoreBinding • SEMplR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate binding propensity for all SEM motifs and genomic positions provided — scoreBinding"><meta name="description" content="Calculate binding propensity for all SEM motifs and
genomic positions provided"><meta property="og:description" content="Calculate binding propensity for all SEM motifs and
genomic positions provided"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SEMplR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SEMplR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/grkenney/SEMplR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate binding propensity for all SEM motifs and genomic positions provided</h1>
      <small class="dont-index">Source: <a href="https://github.com/grkenney/SEMplR/blob/devel/R/scoreBinding.R" class="external-link"><code>R/scoreBinding.R</code></a></small>
      <div class="d-none name"><code>scoreBinding.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate binding propensity for all SEM motifs and
genomic positions provided</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scoreBinding</span><span class="op">(</span><span class="va">x</span>, <span class="va">sem</span>, <span class="va">genome</span>, nFlank <span class="op">=</span> <span class="cn">NULL</span>, seqId <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p><code>GRanges</code> object or a vector of DNA sequences</p></dd>


<dt id="arg-sem">sem<a class="anchor" aria-label="anchor" href="#arg-sem"></a></dt>
<dd><p>A <code>SNPEffectMatrix</code> or <code>SNPEffectMatrixCollection</code> object</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>A <code>BSgenome</code> object for the genome build to use. ie.
<code><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">BSgenome.Hsapiens.UCSC.hg19::Hsapiens</a></code>. Required if providing a GRanges
object. Ignored if providing a vector of sequences.</p></dd>


<dt id="arg-nflank">nFlank<a class="anchor" aria-label="anchor" href="#arg-nflank"></a></dt>
<dd><p>Number of flanking nucleotides to add to provided range. By
default will add flank equal to the length of the longest motif. Ignored
if providing a vector of sequences.</p></dd>


<dt id="arg-seqid">seqId<a class="anchor" aria-label="anchor" href="#arg-seqid"></a></dt>
<dd><p>Column in <code>GRanges</code> object to use for unique id.
By default, ids will be generated from the <code>seqnames</code> and <code>ranges.</code>
Ignored if not providing a <code>GRanges</code> object.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If a <code>GRanges</code> object is provided, return a <code>SEMplScores</code> object.
If a list of sequences is provided, just return the scoring table</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load SEMs</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create a GRanges object</span></span></span>
<span class="r-in"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    seqnames <span class="op">=</span> <span class="st">"chr12"</span>,</span></span>
<span class="r-in"><span>    ranges <span class="op">=</span> <span class="fl">94136009</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># calculate binding propensity</span></span></span>
<span class="r-in"><span><span class="fu">scoreBinding</span><span class="op">(</span><span class="va">gr</span>, <span class="va">sc</span>, <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg19/man/package.html" class="external-link">Hsapiens</a></span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class SEMplScores</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ranges(1): chr12:94136009</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> semData(13): transcription_factor, ensembl_id ... PWM_source, SEM_KEY</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> scores(223):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               seqId                SEM      score  scoreNorm index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              &lt;char&gt;             &lt;char&gt;      &lt;num&gt;      &lt;num&gt; &lt;int&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1: chr12:94136009 AP2B_HUMAN.SK-N-SH  -1.689754 -0.3068238    15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2: chr12:94136009 ARNT_HUMAN.GM12878  -6.892799 -0.9693833    17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3: chr12:94136009    ATF1_HUMAN.K562  -7.079925 -0.9420095    16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4: chr12:94136009   ATF2_HUMAN.HepG2  -4.890126 -0.9098440    16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   5: chr12:94136009   ATF3_HUMAN.HepG2  -8.605675 -0.9885365    14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ---                                                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 219: chr12:94136009  ZBT7A_HUMAN.HepG2  -1.859506 -0.6349682    12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 220: chr12:94136009    ZFX_HUMAN.HepG2  -1.459472 -0.5682106    19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 221: chr12:94136009  ZN281_HUMAN.HepG2  -4.347612 -0.9355197     8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 222: chr12:94136009 ZNF18_HUMAN.HEK293  -5.410220 -0.9264060    15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 223: chr12:94136009   ZSCAN4_secondary -15.439087 -0.9998307    15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                &lt;char&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1:       GCTTTGAGGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2:        TTTGAGGCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3:      CTTTGAGGCAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4:      CTTTGAGGCAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   5:      GGCTTTGAGGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ---                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 219:        AAGGCTTTG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 220:       TGAGGCATCT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 221:  GGAGAAGGCTTTGAG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 222:     GCTTTGAGGCAT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 223: GCTTTGAGGCATCTGC</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/grkenney" class="external-link">Grace Kenney</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

