<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SEMplR Vignette â€¢ SEMplR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="SEMplR Vignette">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SEMplR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/SEMplR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/grkenney/SEMplR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SEMplR Vignette</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/grkenney/SEMplR/blob/devel/vignettes/SEMplR.Rmd" class="external-link"><code>vignettes/SEMplR.Rmd</code></a></small>
      <div class="d-none name"><code>SEMplR.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grkenney/SEMplR" class="external-link">SEMplR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="snp-effect-matrices">SNP Effect Matrices<a class="anchor" aria-label="anchor" href="#snp-effect-matrices"></a>
</h2>
<p>SEMplR uses SNP Effect Matrices (SEMs) to score potential motifs.
These are matrices contain binding affinity scores and have rows equal
to the length of the motif and a column for each nucleotide. SEMs are
produced by SEMpl, but a default set of 223 are included with this
package in the <code>sc</code> data object. A full list of the
transcription factors included in this default set can be found <a href="github.com/grkenney/SEMplR/blob/main/vignettes/sempl_metadata.csv">here</a>
or by running <code>semData(sc)</code>.</p>
<p>The default SEM collection can be loaded with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span></span></code></pre></div>
<p>Printing the <code>SNPEffectMatrixCollection</code>, we can see that
we have 223 SEMs and 13 meta data features for each.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc</span></span>
<span><span class="co">#&gt; An object of class SNPEffectMatrixCollection</span></span>
<span><span class="co">#&gt; sems(223): AP2B_HUMAN.SK-N-SH, ARNT_HUMAN.GM12878 ... ZNF18_HUMAN.HEK293, ZSCAN4_secondary</span></span>
<span><span class="co">#&gt; semData(13): transcription_factor, ensembl_id ... PWM_source, SEM_KEY</span></span></code></pre></div>
<p>We can access all SEMs in collection the with the function
<code><a href="../reference/sems.html">sems()</a></code> or some subset of matrices by specifying a vector of
semIds in the <code>semId</code> parameter.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sems.html">sems</a></span><span class="op">(</span><span class="va">sc</span>, semId <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AP2B_HUMAN.SK-N-SH"</span>, <span class="st">"ZSCAN4_secondary"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`AP2B_HUMAN.SK-N-SH`</span></span>
<span><span class="co">#&gt; An object of class SNPEffectMatrix</span></span>
<span><span class="co">#&gt; semId:  AP2B_HUMAN.SK-N-SH</span></span>
<span><span class="co">#&gt; baseline:  -1.161048</span></span>
<span><span class="co">#&gt; sem:</span></span>
<span><span class="co">#&gt;             A         C         G         T</span></span>
<span><span class="co">#&gt;         &lt;num&gt;     &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1: -1.093659 -0.228215 -0.035353 -0.949757</span></span>
<span><span class="co">#&gt;  2: -1.198571  0.019236 -1.057251 -1.145949</span></span>
<span><span class="co">#&gt;  3: -1.021307  0.008023 -1.273623 -0.746918</span></span>
<span><span class="co">#&gt;  4: -0.808380 -0.281746 -0.849156  0.093359</span></span>
<span><span class="co">#&gt;  5: -0.319849 -0.071797 -0.058358 -0.085237</span></span>
<span><span class="co">#&gt;  6:  0.119797 -0.640756 -0.336054 -0.787809</span></span>
<span><span class="co">#&gt;  7: -0.496324 -1.111405  0.034943 -0.894114</span></span>
<span><span class="co">#&gt;  8: -1.069881 -1.178852  0.005380 -1.079589</span></span>
<span><span class="co">#&gt;  9: -0.957910 -0.095025 -0.061232 -1.025970</span></span>
<span><span class="co">#&gt; 10: -0.147762 -0.046611 -0.308033 -0.113474</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ZSCAN4_secondary</span></span>
<span><span class="co">#&gt; An object of class SNPEffectMatrix</span></span>
<span><span class="co">#&gt; semId:  ZSCAN4_secondary</span></span>
<span><span class="co">#&gt; baseline:  -2.911018</span></span>
<span><span class="co">#&gt; sem:</span></span>
<span><span class="co">#&gt;             A          C         G        T</span></span>
<span><span class="co">#&gt;         &lt;num&gt;      &lt;num&gt;     &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:  0.339681 -1.6258400 -0.852570 -1.40307</span></span>
<span><span class="co">#&gt;  2: -1.858890  0.2245480 -1.803160 -1.10145</span></span>
<span><span class="co">#&gt;  3:  0.259177 -1.8665800 -0.744934 -1.62341</span></span>
<span><span class="co">#&gt;  4: -2.077230  0.1970890 -1.971940 -1.21187</span></span>
<span><span class="co">#&gt;  5:  0.198878 -2.2059500 -0.760569 -1.76139</span></span>
<span><span class="co">#&gt;  6: -2.321170  0.1354730 -2.300640 -1.39237</span></span>
<span><span class="co">#&gt;  7:  0.157553 -2.4685400 -0.741112 -1.84194</span></span>
<span><span class="co">#&gt;  8: -2.482910  0.0893564 -2.406690 -1.56081</span></span>
<span><span class="co">#&gt;  9:  0.136749 -2.4335100 -0.704496 -1.94128</span></span>
<span><span class="co">#&gt; 10: -2.352060  0.1257340 -2.211630 -1.60719</span></span>
<span><span class="co">#&gt; 11:  0.187489 -2.1923800 -0.641008 -1.76197</span></span>
<span><span class="co">#&gt; 12: -2.109030  0.1677660 -1.974360 -1.51917</span></span>
<span><span class="co">#&gt; 13:  0.240220 -1.9354700 -0.655348 -1.54698</span></span>
<span><span class="co">#&gt; 14: -1.748380  0.2142480 -1.693530 -1.38530</span></span>
<span><span class="co">#&gt; 15:  0.289375 -1.6708800 -0.698449 -1.42560</span></span>
<span><span class="co">#&gt; 16: -1.483480  0.2650430 -1.541780 -1.31953</span></span></code></pre></div>
<p>We can view the SEM meta data slot with the <code>semData</code>
function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/semData.html">semData</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span></span>
<span><span class="co">#&gt; Key: &lt;SEM_KEY&gt;</span></span>
<span><span class="co">#&gt;      transcription_factor      ensembl_id ebi_complex_ac uniprot_ac</span></span>
<span><span class="co">#&gt;                    &lt;char&gt;          &lt;char&gt;         &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1:               TFAP2B ENSG00000008196                    Q92481</span></span>
<span><span class="co">#&gt;   2:                 ARNT ENSG00000143437                    P27540</span></span>
<span><span class="co">#&gt;   3:                 ATF1 ENSG00000123268                    P18846</span></span>
<span><span class="co">#&gt;   4:                 ATF2 ENSG00000115966                    P15336</span></span>
<span><span class="co">#&gt;   5:                 ATF3 ENSG00000162772                    P18847</span></span>
<span><span class="co">#&gt;  ---                                                               </span></span>
<span><span class="co">#&gt; 219:               ZBTB7A ENSG00000178951                    O95365</span></span>
<span><span class="co">#&gt; 220:                  ZFX ENSG00000005889                    P17010</span></span>
<span><span class="co">#&gt; 221:               ZNF281 ENSG00000162702                    Q9Y2X9</span></span>
<span><span class="co">#&gt; 222:                ZNF18 ENSG00000154957                    P17022</span></span>
<span><span class="co">#&gt; 223:               ZSCAN4 ENSG00000180532                    Q8NAM6</span></span>
<span><span class="co">#&gt;                     PWM_id                    SEM SEM_baseline cell_type</span></span>
<span><span class="co">#&gt;                     &lt;char&gt;                 &lt;char&gt;        &lt;num&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1:  AP2B_HUMAN.H11MO.0.B AP2B_HUMAN.SK-N-SH.sem   -1.1610478   SK-N-SH</span></span>
<span><span class="co">#&gt;   2:  ARNT_HUMAN.H11MO.0.B ARNT_HUMAN.GM12878.sem   -1.8632604   GM12878</span></span>
<span><span class="co">#&gt;   3:  ATF1_HUMAN.H11MO.0.B    ATF1_HUMAN.K562.sem   -2.9718855      K562</span></span>
<span><span class="co">#&gt;   4:  ATF2_HUMAN.H11MO.0.B   ATF2_HUMAN.HepG2.sem   -1.4186930     HepG2</span></span>
<span><span class="co">#&gt;   5:  ATF3_HUMAN.H11MO.0.A   ATF3_HUMAN.HepG2.sem   -2.1588687     HepG2</span></span>
<span><span class="co">#&gt;  ---                                                                    </span></span>
<span><span class="co">#&gt; 219: ZBT7A_HUMAN.H11MO.0.A  ZBT7A_HUMAN.HepG2.sem   -0.4056000     HepG2</span></span>
<span><span class="co">#&gt; 220:   ZFX_HUMAN.H11MO.0.A    ZFX_HUMAN.HepG2.sem   -0.2478717     HepG2</span></span>
<span><span class="co">#&gt; 221: ZN281_HUMAN.H11MO.0.A  ZN281_HUMAN.HepG2.sem   -0.3926150     HepG2</span></span>
<span><span class="co">#&gt; 222: ZNF18_HUMAN.H11MO.0.C ZNF18_HUMAN.HEK293.sem   -1.6459529    HEK293</span></span>
<span><span class="co">#&gt; 223:  Zscan4_pwm_secondary   ZSCAN4_secondary.sem   -2.9110180    HEK293</span></span>
<span><span class="co">#&gt;      neg_log10_pval chip_ENCODE_accession dnase_ENCODE_accession  PWM_source</span></span>
<span><span class="co">#&gt;               &lt;num&gt;                &lt;char&gt;                 &lt;char&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1:       6.476702           ENCFF303ORH            ENCFF752OZB HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   2:      12.081160           ENCFF538DNI            ENCFF759OLD HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   3:      56.056680           ENCFF885JPB            ENCFF274YGF HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   4:      27.821210           ENCFF899JAX            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   5:      14.618530           ENCFF197JBM            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt;  ---                                                                        </span></span>
<span><span class="co">#&gt; 219:       3.684286           ENCFF439DRT            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 220:       3.443155           ENCFF329DCV            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 221:       3.428189           ENCFF948PYK            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 222:       3.793244           ENCFF888BHA            ENCFF285OXK HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 223:      11.973690           ENCFF485WHZ            ENCFF910QHN    UNIPROBE</span></span>
<span><span class="co">#&gt;                 SEM_KEY</span></span>
<span><span class="co">#&gt;                  &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1: AP2B_HUMAN.SK-N-SH</span></span>
<span><span class="co">#&gt;   2: ARNT_HUMAN.GM12878</span></span>
<span><span class="co">#&gt;   3:    ATF1_HUMAN.K562</span></span>
<span><span class="co">#&gt;   4:   ATF2_HUMAN.HepG2</span></span>
<span><span class="co">#&gt;   5:   ATF3_HUMAN.HepG2</span></span>
<span><span class="co">#&gt;  ---                   </span></span>
<span><span class="co">#&gt; 219:  ZBT7A_HUMAN.HepG2</span></span>
<span><span class="co">#&gt; 220:    ZFX_HUMAN.HepG2</span></span>
<span><span class="co">#&gt; 221:  ZN281_HUMAN.HepG2</span></span>
<span><span class="co">#&gt; 222: ZNF18_HUMAN.HEK293</span></span>
<span><span class="co">#&gt; 223:   ZSCAN4_secondary</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scoring-binding">Scoring Binding<a class="anchor" aria-label="anchor" href="#scoring-binding"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSEM.html">plotSEM</a></span><span class="op">(</span><span class="va">sc</span>,</span>
<span>    motif <span class="op">=</span> <span class="st">"AP2B_HUMAN.SK-N-SH"</span>,</span>
<span>    motifSeq <span class="op">=</span> <span class="st">"GCTTTGAGGC"</span>, highlight <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="SEMplR_files/figure-html/unnamed-chunk-6-1.png" width="480"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSEM.html">plotSEM</a></span><span class="op">(</span><span class="va">sc</span>,</span>
<span>    motif <span class="op">=</span> <span class="st">"AP2B_HUMAN.SK-N-SH"</span>,</span>
<span>    motifSeq <span class="op">=</span> <span class="st">"GCTTTCAGGC"</span>, highlight <span class="op">=</span> <span class="fl">1</span>, hcol <span class="op">=</span> <span class="st">"red"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="SEMplR_files/figure-html/unnamed-chunk-6-2.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="scoring-variants">Scoring Variants<a class="anchor" aria-label="anchor" href="#scoring-variants"></a>
</h2>
<div class="section level3">
<h3 id="prepare-inputs">Prepare inputs<a class="anchor" aria-label="anchor" href="#prepare-inputs"></a>
</h3>
<p>First, we will define the variants we want to score. Variants can be
supplied to SEMplR as either a <code>VRanges</code> or a
<code>GRanges</code> object. If using <code>VRanges</code> alleles must
be stored in the <code>ref</code> and <code>alt</code> parameters. If
using <code>GRanges</code> the alleles should be stored in seperate
metadata columns.</p>
<p>Insertions and deletions should be specified by an empty character
vector in the appropriate allele.</p>
<p>Optionally, an <code>id</code> column can be defined in the object
metadata to be used as a unique identifier for each variant.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">VRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr12"</span>, <span class="st">"chr19"</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">94136009</span>, <span class="fl">10640062</span><span class="op">)</span>,</span>
<span>    ref <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G"</span>, <span class="st">"T"</span><span class="op">)</span>, alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="st">"A"</span><span class="op">)</span>,</span>
<span>    id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"variant1"</span>, <span class="st">"variant2"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scoring">Scoring<a class="anchor" aria-label="anchor" href="#scoring"></a>
</h3>
<p>The <code>scoreVariants</code> function scores each allele of each
variant against every SEM provided. From the resulting data object, we
can see that we scored 2 variants, the SEM meta data stored in the
object has 13 fields, and we have 446 rows in the <code>scores</code>
results table.</p>
<p>There is a row in the <code>scores</code> table for each variant/SEM
combination. Here we scored 2 variants x 223 SEMs to get 446 rows.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sempl_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scoreVariants.html">scoreVariants</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">vr</span>,</span>
<span>    sem <span class="op">=</span> <span class="va">sc</span>,</span>
<span>    genome <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span>,</span>
<span>    varId <span class="op">=</span> <span class="st">"id"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Allele 'G' or 'C' does not match reference sequence 'T' in</span></span>
<span><span class="co">#&gt; chr12:94136009</span></span>
<span></span>
<span><span class="va">sempl_results</span></span>
<span><span class="co">#&gt; An object of class SEMplScores</span></span>
<span><span class="co">#&gt; ranges(2): variant1, variant2</span></span>
<span><span class="co">#&gt; semData(13): transcription_factor, ensembl_id ... PWM_source, SEM_KEY</span></span>
<span><span class="co">#&gt; scores(446):</span></span>
<span><span class="co">#&gt;         varId              semId           refSeq           altSeq   refScore</span></span>
<span><span class="co">#&gt;        &lt;char&gt;             &lt;char&gt;           &lt;char&gt;           &lt;char&gt;      &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1: variant1 AP2B_HUMAN.SK-N-SH       GTTTAACAAT       TCTTTAACAA  -5.287583</span></span>
<span><span class="co">#&gt;   2: variant1 ARNT_HUMAN.GM12878        TTGTTTAAC        TGTTGTTCT  -5.048425</span></span>
<span><span class="co">#&gt;   3: variant1    ATF1_HUMAN.K562      ACTTGTTGTTG      CTTTAACAATA  -7.736531</span></span>
<span><span class="co">#&gt;   4: variant1   ATF2_HUMAN.HepG2      CTTGTTGTTGT      CTTGTTGTTCT  -5.049124</span></span>
<span><span class="co">#&gt;   5: variant1   ATF3_HUMAN.HepG2      TGTTGTTTAAC      TGTTCTTTAAC  -9.214165</span></span>
<span><span class="co">#&gt;  ---                                                                         </span></span>
<span><span class="co">#&gt; 442: variant2  ZBT7A_HUMAN.HepG2        TTTAATCCT        TATAATCCT  -2.821984</span></span>
<span><span class="co">#&gt; 443: variant2    ZFX_HUMAN.HepG2       TTAATCCTCT       ATAATCCTCT  -1.407772</span></span>
<span><span class="co">#&gt; 444: variant2  ZN281_HUMAN.HepG2  CTTGGGCAAATTATT  CTTGGGCAAATTATA  -5.109413</span></span>
<span><span class="co">#&gt; 445: variant2 ZNF18_HUMAN.HEK293     TATTTAATCCTC     TATATAATCCTC  -6.516790</span></span>
<span><span class="co">#&gt; 446: variant2   ZSCAN4_secondary TCTTGGGCAAATTATT TTATATAATCCTCTAA -17.314249</span></span>
<span><span class="co">#&gt;        altScore    refNorm    altNorm refVarIndex altVarIndex</span></span>
<span><span class="co">#&gt;           &lt;num&gt;      &lt;num&gt;      &lt;num&gt;       &lt;int&gt;       &lt;int&gt;</span></span>
<span><span class="co">#&gt;   1:  -4.330368 -0.9427482 -0.8888423          20          19</span></span>
<span><span class="co">#&gt;   2:  -5.948900 -0.8900564 -0.9411021          18          13</span></span>
<span><span class="co">#&gt;   3:  -8.123991 -0.9632126 -0.9718770          10          20</span></span>
<span><span class="co">#&gt;   4:  -5.397180 -0.9192521 -0.9365610          11          11</span></span>
<span><span class="co">#&gt;   5:  -9.785488 -0.9924813 -0.9949399          16          16</span></span>
<span><span class="co">#&gt;  ---                                                         </span></span>
<span><span class="co">#&gt; 442:  -2.502507 -0.8126749 -0.7662411          19          19</span></span>
<span><span class="co">#&gt; 443:  -1.508060 -0.5524565 -0.5825105          20          20</span></span>
<span><span class="co">#&gt; 444:  -5.061910 -0.9619721 -0.9606991           6           6</span></span>
<span><span class="co">#&gt; 445:  -6.314741 -0.9658232 -0.9606853          17          17</span></span>
<span><span class="co">#&gt; 446: -17.018032 -0.9999538 -0.9999433           5          16</span></span></code></pre></div>
<p>The scoring results has 10 columns:</p>
<ul>
<li><p><strong>varId</strong>: unique id of the variant as defined in
the <code>id</code> column of the input <code>GRanges</code> or
<code>VRanges</code> object. If not defined, a custom unique identifier
is generated in the format
[seqname]:[range]:[ref_allele]&gt;[alt_allele]</p></li>
<li><p><strong>semId</strong>: the unique identifier of the SEM</p></li>
<li><p><strong>refSeq</strong> and <strong>altSeq</strong>: the
optimally scored sequence for the reference and alternative alleles
respectively</p></li>
<li><p><strong>refScore</strong> and <strong>altScore</strong>: the raw
(unnormalized) binding affinity score for the reference and alternative
alleles respectively</p></li>
<li><p><strong>refNorm</strong> and <strong>altNorm</strong>: the
normalized binding affinity score for the reference and alternative
alleles respectively</p></li>
<li><p><strong>refVarIndex</strong> and <strong>altVarIndex</strong>:
The position of the frameâ€™s starting indeces in the scored
sequence</p></li>
</ul>
<p>There are accessor functions to isolate each slot of the resulting
data object:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access the variants slot</span></span>
<span><span class="fu"><a href="../reference/getRanges.html">getRanges</a></span><span class="op">(</span><span class="va">sempl_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; VRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames    ranges strand         ref              alt     totalDepth</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; &lt;character&gt; &lt;characterOrRle&gt; &lt;integerOrRle&gt;</span></span>
<span><span class="co">#&gt;   [1]    chr12  94136009      *           G                C           &lt;NA&gt;</span></span>
<span><span class="co">#&gt;   [2]    chr19  10640062      *           T                A           &lt;NA&gt;</span></span>
<span><span class="co">#&gt;             refDepth       altDepth   sampleNames softFilterMatrix |</span></span>
<span><span class="co">#&gt;       &lt;integerOrRle&gt; &lt;integerOrRle&gt; &lt;factorOrRle&gt;         &lt;matrix&gt; |</span></span>
<span><span class="co">#&gt;   [1]           &lt;NA&gt;           &lt;NA&gt;          &lt;NA&gt;                  |</span></span>
<span><span class="co">#&gt;   [2]           &lt;NA&gt;           &lt;NA&gt;          &lt;NA&gt;                  |</span></span>
<span><span class="co">#&gt;                id                ref_seq                alt_seq</span></span>
<span><span class="co">#&gt;       &lt;character&gt;            &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt;   [1]    variant1 CTAATAAATACTTGTTGTTG.. CTAATAAATACTTGTTGTTC..</span></span>
<span><span class="co">#&gt;   [2]    variant2 GTAATCTTGGGCAAATTATT.. GTAATCTTGGGCAAATTATA..</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">#&gt;   hardFilters: NULL</span></span>
<span></span>
<span><span class="co"># access the semData slot</span></span>
<span><span class="fu"><a href="../reference/semData.html">semData</a></span><span class="op">(</span><span class="va">sempl_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; Key: &lt;SEM_KEY&gt;</span></span>
<span><span class="co">#&gt;      transcription_factor      ensembl_id ebi_complex_ac uniprot_ac</span></span>
<span><span class="co">#&gt;                    &lt;char&gt;          &lt;char&gt;         &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1:               TFAP2B ENSG00000008196                    Q92481</span></span>
<span><span class="co">#&gt;   2:                 ARNT ENSG00000143437                    P27540</span></span>
<span><span class="co">#&gt;   3:                 ATF1 ENSG00000123268                    P18846</span></span>
<span><span class="co">#&gt;   4:                 ATF2 ENSG00000115966                    P15336</span></span>
<span><span class="co">#&gt;   5:                 ATF3 ENSG00000162772                    P18847</span></span>
<span><span class="co">#&gt;  ---                                                               </span></span>
<span><span class="co">#&gt; 219:               ZBTB7A ENSG00000178951                    O95365</span></span>
<span><span class="co">#&gt; 220:                  ZFX ENSG00000005889                    P17010</span></span>
<span><span class="co">#&gt; 221:               ZNF281 ENSG00000162702                    Q9Y2X9</span></span>
<span><span class="co">#&gt; 222:                ZNF18 ENSG00000154957                    P17022</span></span>
<span><span class="co">#&gt; 223:               ZSCAN4 ENSG00000180532                    Q8NAM6</span></span>
<span><span class="co">#&gt;                     PWM_id                    SEM SEM_baseline cell_type</span></span>
<span><span class="co">#&gt;                     &lt;char&gt;                 &lt;char&gt;        &lt;num&gt;    &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1:  AP2B_HUMAN.H11MO.0.B AP2B_HUMAN.SK-N-SH.sem   -1.1610478   SK-N-SH</span></span>
<span><span class="co">#&gt;   2:  ARNT_HUMAN.H11MO.0.B ARNT_HUMAN.GM12878.sem   -1.8632604   GM12878</span></span>
<span><span class="co">#&gt;   3:  ATF1_HUMAN.H11MO.0.B    ATF1_HUMAN.K562.sem   -2.9718855      K562</span></span>
<span><span class="co">#&gt;   4:  ATF2_HUMAN.H11MO.0.B   ATF2_HUMAN.HepG2.sem   -1.4186930     HepG2</span></span>
<span><span class="co">#&gt;   5:  ATF3_HUMAN.H11MO.0.A   ATF3_HUMAN.HepG2.sem   -2.1588687     HepG2</span></span>
<span><span class="co">#&gt;  ---                                                                    </span></span>
<span><span class="co">#&gt; 219: ZBT7A_HUMAN.H11MO.0.A  ZBT7A_HUMAN.HepG2.sem   -0.4056000     HepG2</span></span>
<span><span class="co">#&gt; 220:   ZFX_HUMAN.H11MO.0.A    ZFX_HUMAN.HepG2.sem   -0.2478717     HepG2</span></span>
<span><span class="co">#&gt; 221: ZN281_HUMAN.H11MO.0.A  ZN281_HUMAN.HepG2.sem   -0.3926150     HepG2</span></span>
<span><span class="co">#&gt; 222: ZNF18_HUMAN.H11MO.0.C ZNF18_HUMAN.HEK293.sem   -1.6459529    HEK293</span></span>
<span><span class="co">#&gt; 223:  Zscan4_pwm_secondary   ZSCAN4_secondary.sem   -2.9110180    HEK293</span></span>
<span><span class="co">#&gt;      neg_log10_pval chip_ENCODE_accession dnase_ENCODE_accession  PWM_source</span></span>
<span><span class="co">#&gt;               &lt;num&gt;                &lt;char&gt;                 &lt;char&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1:       6.476702           ENCFF303ORH            ENCFF752OZB HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   2:      12.081160           ENCFF538DNI            ENCFF759OLD HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   3:      56.056680           ENCFF885JPB            ENCFF274YGF HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   4:      27.821210           ENCFF899JAX            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt;   5:      14.618530           ENCFF197JBM            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt;  ---                                                                        </span></span>
<span><span class="co">#&gt; 219:       3.684286           ENCFF439DRT            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 220:       3.443155           ENCFF329DCV            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 221:       3.428189           ENCFF948PYK            ENCFF897NME HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 222:       3.793244           ENCFF888BHA            ENCFF285OXK HOCOMOCOv11</span></span>
<span><span class="co">#&gt; 223:      11.973690           ENCFF485WHZ            ENCFF910QHN    UNIPROBE</span></span>
<span><span class="co">#&gt;                 SEM_KEY</span></span>
<span><span class="co">#&gt;                  &lt;char&gt;</span></span>
<span><span class="co">#&gt;   1: AP2B_HUMAN.SK-N-SH</span></span>
<span><span class="co">#&gt;   2: ARNT_HUMAN.GM12878</span></span>
<span><span class="co">#&gt;   3:    ATF1_HUMAN.K562</span></span>
<span><span class="co">#&gt;   4:   ATF2_HUMAN.HepG2</span></span>
<span><span class="co">#&gt;   5:   ATF3_HUMAN.HepG2</span></span>
<span><span class="co">#&gt;  ---                   </span></span>
<span><span class="co">#&gt; 219:  ZBT7A_HUMAN.HepG2</span></span>
<span><span class="co">#&gt; 220:    ZFX_HUMAN.HepG2</span></span>
<span><span class="co">#&gt; 221:  ZN281_HUMAN.HepG2</span></span>
<span><span class="co">#&gt; 222: ZNF18_HUMAN.HEK293</span></span>
<span><span class="co">#&gt; 223:   ZSCAN4_secondary</span></span>
<span></span>
<span><span class="co"># access the scores slot</span></span>
<span><span class="fu"><a href="../reference/scores.html">scores</a></span><span class="op">(</span><span class="va">sempl_results</span><span class="op">)</span></span>
<span><span class="co">#&gt;         varId              semId           refSeq           altSeq   refScore</span></span>
<span><span class="co">#&gt;        &lt;char&gt;             &lt;char&gt;           &lt;char&gt;           &lt;char&gt;      &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1: variant1 AP2B_HUMAN.SK-N-SH       GTTTAACAAT       TCTTTAACAA  -5.287583</span></span>
<span><span class="co">#&gt;   2: variant1 ARNT_HUMAN.GM12878        TTGTTTAAC        TGTTGTTCT  -5.048425</span></span>
<span><span class="co">#&gt;   3: variant1    ATF1_HUMAN.K562      ACTTGTTGTTG      CTTTAACAATA  -7.736531</span></span>
<span><span class="co">#&gt;   4: variant1   ATF2_HUMAN.HepG2      CTTGTTGTTGT      CTTGTTGTTCT  -5.049124</span></span>
<span><span class="co">#&gt;   5: variant1   ATF3_HUMAN.HepG2      TGTTGTTTAAC      TGTTCTTTAAC  -9.214165</span></span>
<span><span class="co">#&gt;  ---                                                                         </span></span>
<span><span class="co">#&gt; 442: variant2  ZBT7A_HUMAN.HepG2        TTTAATCCT        TATAATCCT  -2.821984</span></span>
<span><span class="co">#&gt; 443: variant2    ZFX_HUMAN.HepG2       TTAATCCTCT       ATAATCCTCT  -1.407772</span></span>
<span><span class="co">#&gt; 444: variant2  ZN281_HUMAN.HepG2  CTTGGGCAAATTATT  CTTGGGCAAATTATA  -5.109413</span></span>
<span><span class="co">#&gt; 445: variant2 ZNF18_HUMAN.HEK293     TATTTAATCCTC     TATATAATCCTC  -6.516790</span></span>
<span><span class="co">#&gt; 446: variant2   ZSCAN4_secondary TCTTGGGCAAATTATT TTATATAATCCTCTAA -17.314249</span></span>
<span><span class="co">#&gt;        altScore    refNorm    altNorm refVarIndex altVarIndex</span></span>
<span><span class="co">#&gt;           &lt;num&gt;      &lt;num&gt;      &lt;num&gt;       &lt;int&gt;       &lt;int&gt;</span></span>
<span><span class="co">#&gt;   1:  -4.330368 -0.9427482 -0.8888423          20          19</span></span>
<span><span class="co">#&gt;   2:  -5.948900 -0.8900564 -0.9411021          18          13</span></span>
<span><span class="co">#&gt;   3:  -8.123991 -0.9632126 -0.9718770          10          20</span></span>
<span><span class="co">#&gt;   4:  -5.397180 -0.9192521 -0.9365610          11          11</span></span>
<span><span class="co">#&gt;   5:  -9.785488 -0.9924813 -0.9949399          16          16</span></span>
<span><span class="co">#&gt;  ---                                                         </span></span>
<span><span class="co">#&gt; 442:  -2.502507 -0.8126749 -0.7662411          19          19</span></span>
<span><span class="co">#&gt; 443:  -1.508060 -0.5524565 -0.5825105          20          20</span></span>
<span><span class="co">#&gt; 444:  -5.061910 -0.9619721 -0.9606991           6           6</span></span>
<span><span class="co">#&gt; 445:  -6.314741 -0.9658232 -0.9606853          17          17</span></span>
<span><span class="co">#&gt; 446: -17.018032 -0.9999538 -0.9999433           5          16</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSemMotifs.html">plotSemMotifs</a></span><span class="op">(</span><span class="va">sempl_results</span>,</span>
<span>    variant <span class="op">=</span> <span class="st">"variant2"</span>,</span>
<span>    label <span class="op">=</span> <span class="st">"transcription_factor"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="SEMplR_files/figure-html/unnamed-chunk-10-1.png" width="480"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSemVariants.html">plotSemVariants</a></span><span class="op">(</span><span class="va">sempl_results</span>, sem <span class="op">=</span> <span class="st">"AP2B_HUMAN.SK-N-SH"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SEMplR_files/figure-html/unnamed-chunk-11-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="enrichment">Enrichment<a class="anchor" aria-label="anchor" href="#enrichment"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loading data</span></span>
<span><span class="va">caqtl_file</span> <span class="op">&lt;-</span> <span class="st">"caQTL_variants_overlappingPeaks_LD-r2-0.8_withLead.bed"</span></span>
<span><span class="va">caqtls</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">caqtl_file</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">proxy_ID_split</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">caqtls</span><span class="op">$</span><span class="va">proxy_ID</span>, <span class="st">":"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># formatting position and alleles</span></span>
<span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">proxy_ID_split</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">proxy_ID_split</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ref_allele</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">proxy_ID_split</span>, <span class="va">`[[`</span>, <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">alt_allele</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">proxy_ID_split</span>, <span class="va">`[[`</span>, <span class="fl">4</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># building variant annotation obj</span></span>
<span><span class="va">vr_e</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VRanges-class.html" class="external-link">VRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="va">chr</span>,</span>
<span>    ranges <span class="op">=</span> <span class="va">pos</span>,</span>
<span>    ref <span class="op">=</span> <span class="va">ref_allele</span>,</span>
<span>    alt <span class="op">=</span> <span class="va">alt_allele</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># score variants</span></span>
<span><span class="va">var_scores_e</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scoreVariants.html">scoreVariants</a></span><span class="op">(</span></span>
<span>    vr <span class="op">=</span> <span class="va">vr_e</span>,</span>
<span>    semList <span class="op">=</span> <span class="va">sc</span>,</span>
<span>    bs_genome_obj <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSemVariants.html">plotSemVariants</a></span><span class="op">(</span><span class="va">var_scores_e</span>, semId <span class="op">=</span> <span class="st">"MA0151.1"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSemMotifs.html">plotSemMotifs</a></span><span class="op">(</span><span class="va">var_scores_e</span>, <span class="st">"chr1:906982:C&gt;T"</span>, label <span class="op">=</span> <span class="st">"transcription_factor"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="binding-enrichment">Binding Enrichment<a class="anchor" aria-label="anchor" href="#binding-enrichment"></a>
</h2>
<p>Given this is a large set of variants, we might want to check if some
transcription factors are enriched for binding in one allele or the
other.</p>
<p>Here, we will test if any of these transcription factors are bound
more than expected in the reference alleles of this set.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr_e</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="va">chr</span>,</span>
<span>    ranges <span class="op">=</span> <span class="va">pos</span>,</span>
<span>    allele <span class="op">=</span> <span class="va">ref_allele</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrichSEMs.html">enrichSEMs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">gr_e</span>, semList <span class="op">=</span> <span class="va">sc</span>,</span>
<span>    bs_genome_obj <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotEnrich.html">plotEnrich</a></span><span class="op">(</span><span class="va">e</span>, semList <span class="op">=</span> <span class="va">sc</span><span class="op">)</span></span>
<span><span class="va">plt</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extras">Extras<a class="anchor" aria-label="anchor" href="#extras"></a>
</h2>
<div class="section level3">
<h3 id="loading-sems">Loading SEMs<a class="anchor" aria-label="anchor" href="#loading-sems"></a>
</h3>
<p>First, we create a list of file paths to the .sem files we want to
include. We will also load the meta data for each sem in a
<code>data.table</code> object. If meta data is used, all SEMs must be
represented in the meta data table.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sem_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"SEMs/"</span>,</span>
<span>    pattern <span class="op">=</span> <span class="st">".sem"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">metadata_file</span> <span class="op">&lt;-</span> <span class="st">"SEMs/sempl_metadata.csv"</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">metadata_file</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span></span></code></pre></div>
<p>We will load the matrix and meta data for all SEMs in a single
<code>SNPEffectMatrixCollection</code> object. This object has two
slots, one containing a named list of the matrices and a second slot
containing our meta data table with a key column connecting the meta
data to the names of the matrices.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadSEMCollection.html">loadSEMCollection</a></span><span class="op">(</span></span>
<span>    semFiles <span class="op">=</span> <span class="va">sem_files</span>,</span>
<span>    semMetaData <span class="op">=</span> <span class="va">meta</span>,</span>
<span>    semMetaKey <span class="op">=</span> <span class="st">"SEM"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sc</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://sessioninfo.r-lib.org/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  en_US.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    en_US.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2025-09-23</span></span>
<span><span class="co">#&gt;  pandoc   3.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt;  quarto   1.4.555 @ /usr/local/bin/quarto</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package                    </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version  </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  abind                         1.4-8     <span style="color: #555555;">2024-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  AnnotationDbi                 1.66.0    <span style="color: #555555;">2024-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  Biobase                     * 2.64.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  BiocGenerics                * 0.50.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  BiocIO                        1.14.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  BiocParallel                  1.38.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  Biostrings                  * 2.72.1    <span style="color: #555555;">2024-06-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  bit                           4.6.0     <span style="color: #555555;">2025-03-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  bit64                         4.6.0-1   <span style="color: #555555;">2025-01-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  bitops                        1.0-9     <span style="color: #555555;">2024-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  blob                          1.2.4     <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  BSgenome                      1.72.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  BSgenome.Hsapiens.UCSC.hg38   1.4.5     <span style="color: #555555;">2024-10-27</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bslib                         0.9.0     <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  cachem                        1.1.0     <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  cli                           3.6.5     <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  codetools                     0.2-20    <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  crayon                        1.5.3     <span style="color: #555555;">2024-06-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  curl                          7.0.0     <span style="color: #555555;">2025-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  data.table                    1.17.8    <span style="color: #555555;">2025-07-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  DBI                           1.2.3     <span style="color: #555555;">2024-06-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  DelayedArray                  0.30.1    <span style="color: #555555;">2024-05-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  desc                          1.4.3     <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  devtools                      2.4.5     <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  digest                        0.6.37    <span style="color: #555555;">2024-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  dplyr                         1.1.4     <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  ellipsis                      0.3.2     <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  evaluate                      1.0.5     <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  farver                        2.1.2     <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  fastmap                       1.2.0     <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  fs                            1.6.6     <span style="color: #555555;">2025-04-12</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  generics                      0.1.4     <span style="color: #555555;">2025-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb                * 1.40.1    <span style="color: #555555;">2024-05-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData              1.2.12    <span style="color: #555555;">2024-10-27</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomicAlignments             1.40.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  GenomicFeatures               1.56.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  GenomicRanges               * 1.56.2    <span style="color: #555555;">2024-10-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  ggplot2                       4.0.0     <span style="color: #555555;">2025-09-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  ggrepel                       0.9.6     <span style="color: #555555;">2024-09-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  glue                          1.8.0     <span style="color: #555555;">2024-09-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  gtable                        0.3.6     <span style="color: #555555;">2024-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  htmltools                     0.5.8.1   <span style="color: #555555;">2024-04-04</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  htmlwidgets                   1.6.4     <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  httpuv                        1.6.16    <span style="color: #555555;">2025-04-16</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  httr                          1.4.7     <span style="color: #555555;">2023-08-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  IRanges                     * 2.38.1    <span style="color: #555555;">2024-07-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  jquerylib                     0.1.4     <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  jsonlite                      2.0.0     <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  KEGGREST                      1.44.1    <span style="color: #555555;">2024-06-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  knitr                         1.50      <span style="color: #555555;">2025-03-16</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  labeling                      0.4.3     <span style="color: #555555;">2023-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  later                         1.4.4     <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  lattice                       0.22-7    <span style="color: #555555;">2025-04-02</span> <span style="color: #555555;">[3]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  lifecycle                     1.0.4     <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  magrittr                      2.0.4     <span style="color: #555555;">2025-09-12</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  Matrix                        1.7-4     <span style="color: #555555;">2025-08-28</span> <span style="color: #555555;">[3]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  MatrixGenerics              * 1.16.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  matrixStats                 * 1.5.0     <span style="color: #555555;">2025-01-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  memoise                       2.0.1     <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  mime                          0.13      <span style="color: #555555;">2025-03-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  miniUI                        0.1.2     <span style="color: #555555;">2025-04-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pillar                        1.11.1    <span style="color: #555555;">2025-09-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pkgbuild                      1.4.8     <span style="color: #555555;">2025-05-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pkgconfig                     2.0.3     <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pkgdown                       2.1.3     <span style="color: #555555;">2025-05-25</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pkgload                       1.4.0     <span style="color: #555555;">2024-06-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  png                           0.1-8     <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  profvis                       0.4.0     <span style="color: #555555;">2024-09-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  promises                      1.3.3     <span style="color: #555555;">2025-05-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  purrr                         1.1.0     <span style="color: #555555;">2025-07-10</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  R6                            2.6.1     <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  ragg                          1.5.0     <span style="color: #555555;">2025-09-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  RColorBrewer                  1.1-3     <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  Rcpp                          1.1.0     <span style="color: #555555;">2025-07-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  RCurl                         1.98-1.17 <span style="color: #555555;">2025-03-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  remotes                       2.5.0     <span style="color: #555555;">2024-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  restfulr                      0.0.16    <span style="color: #555555;">2025-06-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  rjson                         0.2.23    <span style="color: #555555;">2024-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  rlang                         1.1.6     <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  rmarkdown                     2.29      <span style="color: #555555;">2024-11-04</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  Rsamtools                   * 2.20.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  RSQLite                       2.4.3     <span style="color: #555555;">2025-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  rtracklayer                   1.64.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  S4Arrays                      1.4.1     <span style="color: #555555;">2024-05-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  S4Vectors                   * 0.42.1    <span style="color: #555555;">2024-07-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  S7                            0.2.0     <span style="color: #555555;">2024-11-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  sass                          0.4.10    <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  scales                        1.4.0     <span style="color: #555555;">2025-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  SEMplR                      * 0.99.0    <span style="color: #555555;">2025-09-23</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sessioninfo                   1.2.3     <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  shiny                         1.11.1    <span style="color: #555555;">2025-07-03</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  SparseArray                   1.4.8     <span style="color: #555555;">2024-05-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  SummarizedExperiment        * 1.34.0    <span style="color: #555555;">2024-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  systemfonts                   1.2.3     <span style="color: #555555;">2025-04-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  textshaping                   1.0.3     <span style="color: #555555;">2025-09-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  tibble                        3.3.0     <span style="color: #555555;">2025-06-08</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  tidyselect                    1.2.1     <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  UCSC.utils                    1.0.0     <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  urlchecker                    1.0.1     <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  usethis                       3.2.1     <span style="color: #555555;">2025-09-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  VariantAnnotation           * 1.50.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  vctrs                         0.6.5     <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  withr                         3.0.2     <span style="color: #555555;">2024-10-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  xfun                          0.53      <span style="color: #555555;">2025-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  XML                           3.99-0.19 <span style="color: #555555;">2025-08-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  xtable                        1.8-4     <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  XVector                     * 0.44.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt;  yaml                          2.3.10    <span style="color: #555555;">2024-07-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  zlibbioc                      1.50.0    <span style="color: #555555;">2024-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.1)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /__w/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /usr/local/lib/R/library</span></span></span>
<span><span class="co">#&gt;  <span style="color: #BBBBBB; background-color: #BB0000;">*</span> â”€â”€ Packages attached to the search path.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/grkenney" class="external-link">Grace Kenney</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
